### ISSUE TITLE
Linux builder setup

### ISSUE THREAD

**Issue by finestructure** [SwiftPackageIndex/SwiftPackageIndex-Server#981]
Some permission issue with the container, job: https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1083393503

<img width="456" alt="CleanShot 2021-03-09 at 14 44 07@2x" src="https://user-images.githubusercontent.com/65520/110479695-fa4abc00-80e5-11eb-9ebb-5750c57b2af1.png">

**Comment by finestructure** [comment]
Fixed: https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1083774323

But we've got a follow-up issue

<img width="641" alt="CleanShot 2021-03-09 at 15 26 56@2x" src="https://user-images.githubusercontent.com/65520/110485434-ea35db00-80eb-11eb-850d-480137046a86.png">

**Comment by finestructure** [comment]
Made progress with a shell executer but obviously we still need to swap out the macOS binary for the docker command.

https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1084548714

<img width="507" alt="CleanShot 2021-03-09 at 19 59 00@2x" src="https://user-images.githubusercontent.com/65520/110523078-ee74ef00-8111-11eb-8c55-12604ebac717.png">

**Comment by finestructure** [comment]
We're almost there: https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1086379448

<img width="488" alt="CleanShot 2021-03-10 at 12 05 01@2x" src="https://user-images.githubusercontent.com/65520/110620050-dc8a5f00-8198-11eb-8e18-d9fc9f5b9235.png">

Two remaining issues:

- no logs (see #982)
- the linux builder is failing to send the build report - probably blocked by outbound NSG rules

**Comment by finestructure** [comment]
I'll set up a test for this to see if that gets things in working order. The docker runner is probably fixable but it's getting quite complicated. The blocking issue (apart from the missing logs) is that the runs leave `root` owned artifacts behind that `gitlab-runner` trips over when starting a new run.

**Comment by finestructure** [comment]
Meh, I'm almost done but while fixing up the builder, I realised that we can't do that without breaking builds that require custom images. And we need to support those or else we'd have to install all sorts of dependencies on the runners - which obviously doesn't scale or even work if they're conflicting.

So we need to make the docker build work.

I'm pretty sure I know how to fix the permission issue without too much hassle but I'm not sure we'll get the build logs.

**Comment by finestructure** [comment]
And now I remember why I wanted to use docker to package the linux binary 😅 

<img width="745" alt="CleanShot 2021-03-11 at 11 29 44@2x" src="https://user-images.githubusercontent.com/65520/110773519-1e7cd900-825d-11eb-9287-05e41ff5a097.png">

**Comment by finestructure** [comment]
Ok, it's finally passing: https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1089610859

<img width="858" alt="CleanShot 2021-03-11 at 11 44 14@2x" src="https://user-images.githubusercontent.com/65520/110775424-30f81200-825f-11eb-84e4-07dc931f1265.png">

Notice how it _not_ showing the build logs 😡

**Comment by finestructure** [comment]
Also not showing when running the build manually with the exact same command in the builder dir:

```
root@lb1:/etc/gitlab-runner/builds/DfmHEic9/0/finestructure/swiftpackageindex-builder# ./builder-Linux checkout -c https://github.com/finestructure/Parser.git -r 0.2.0 -f
Clone URL: https://github.com/finestructure/Parser.git
Reference: 0.2.0
Cloned https://github.com/finestructure/Parser.git into ./checkout
root@lb1:/etc/gitlab-runner/builds/DfmHEic9/0/finestructure/swiftpackageindex-builder# ./builder-Linux build -s 5.3 -p linux
Selected platform:         linux
Swift version:             v5_3
Building package at path:  ./checkout
Running build ...
docker run --rm -v "$PWD":/host -w /host swift:5.3 swift build --enable-test-discovery
✅  build complete
root@lb1:/etc/gitlab-runner/builds/DfmHEic9/0/finestructure/swiftpackageindex-builder#
```

The assumption that this is due to the nested docker commands is obviously wrong. So it must be either the OS or the docker version. We're running 19.3 on the Macs and 20.5 on Linux.

It might be that ShellOut is not flushing as expected when running on Linux. I'll have to think of a way to test this without incurring these massively long build-run cycles I have been so far.

**Comment by finestructure** [comment]
For sake of completeness, it's not the docker version (which was unlikely anyway). Downgrading it on the builder doesn't fix it.

**Comment by finestructure** [comment]
It can't be ShellOut, either, because as I mentioned in the other issue simply running docker commands without any of our builder binary shows inconsistent and missing output:

```
root@lb1:/etc/gitlab-runner/builds/DfmHEic9/0/finestructure/swiftpackageindex-builder/checkout# docker run --rm -v "$PWD":/host -w /host swift:5.3 swift build --enable-test-discovery
root@lb1:/etc/gitlab-runner/builds/DfmHEic9/0/finestructure/swiftpackageindex-builder/checkout# docker run --rm -it -v "$PWD":/host -w /host swift:5.3 swift build --enable-test-discovery
[1/1] Merging module Parser
root@lb1:/etc/gitlab-runner/builds/DfmHEic9/0/finestructure/swiftpackageindex-builder/checkout# docker run --rm -it -v "$PWD":/host -w /host swift:5.3 swift build --enable-test-discovery
root@lb1:/etc/gitlab-runner/builds/DfmHEic9/0/finestructure/swiftpackageindex-builder/checkout# docker run --rm -it -v "$PWD":/host -w /host swift:5.3 swift build --enable-test-discovery
root@lb1:/etc/gitlab-runner/builds/DfmHEic9/0/finestructure/swiftpackageindex-builder/checkout# docker run --rm -it -v "$PWD":/host -w /host swift:5.3 swift build --enable-test-discovery
root@lb1:/etc/gitlab-runner/builds/DfmHEic9/0/finestructure/swiftpackageindex-builder/checkout# docker run --rm -v "$PWD":/host -w /host swift:5.3 swift build --enable-test-discovery
[1/1] Merging module Parser
root@lb1:/etc/gitlab-runner/builds/DfmHEic9/0/finestructure/swiftpackageindex-builder/checkout# docker run --rm -it -v "$PWD":/host -w /host swift:5.3 swift build --enable-test-discovery
[1/1] Merging module Parser
```

**Comment by finestructure** [comment]
Turns out we _are_ getting some logs, stderr apparently. If I run the 4.2 Parser build (4.2 is unsupported) I get:

<img width="578" alt="CleanShot 2021-03-11 at 16 30 54@2x" src="https://user-images.githubusercontent.com/65520/110811803-51d45d80-8287-11eb-9785-a912c43d0ee1.png">

https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1090497038

Here's what it looks like in the SPI build detail view:

Live version:

<img width="745" alt="CleanShot 2021-03-11 at 16 38 55@2x" src="https://user-images.githubusercontent.com/65520/110812938-564d4600-8288-11eb-977d-58ae5a18b852.png">

New version:

<img width="729" alt="CleanShot 2021-03-11 at 16 39 46@2x" src="https://user-images.githubusercontent.com/65520/110813090-77159b80-8288-11eb-94a7-5f2273398f1b.png">

stdout/stderr ordering issues (which are unavoidable) aside, all the same contents appears to be there.

Now compare the successful cases to see what's missing:

[Current prod](https://swiftpackageindex.com/builds/0E30BF68-9DE3-4305-9F4C-00FDDED604A3):

<img width="733" alt="CleanShot 2021-03-11 at 16 37 24@2x" src="https://user-images.githubusercontent.com/65520/110812681-1be3a900-8288-11eb-9c67-a091d1182038.png">

Current staging (where I'm pushing the test builds):

<img width="749" alt="CleanShot 2021-03-11 at 16 36 24@2x" src="https://user-images.githubusercontent.com/65520/110812527-fb1b5380-8287-11eb-9ce6-5428cdb5c861.png">

https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1090491994

It looks like the only parts that's missing is the build progress:

```
[1/4] Compiling Parser Match.swift
[2/4] Compiling Parser Parser+zip.swift
[3/4] Compiling Parser CharacterSet+ext.swift
[4/4] Compiling Parser Parser.swift
[5/5] Merging module Parser
```

It'd be nice to have that but I'm beginning to understand what might be happening.

When a terminal's attached, SPM logs the build progress on the same line. In CI it prints a new line for each step. Looks like this might at play when running on Linux and this doesn't get set correctly, resetting everything. That might explain why sometimes the last line shows even when running interactively - it's just messed up.

If that's the only thing we'd be missing (temporarily perhaps) I think we can live with that.

**Comment by finestructure** [comment]
Ok, I've got a working version based on our original docker image based approach that doesn't require any builder setup beyond just docker.

This is what we'll get for succeeded and failed builds:

<img width="746" alt="CleanShot 2021-03-12 at 08 51 58@2x" src="https://user-images.githubusercontent.com/65520/110909392-542eca00-8310-11eb-8c0a-1b35781c02df.png">
<img width="738" alt="CleanShot 2021-03-12 at 08 51 13@2x" src="https://user-images.githubusercontent.com/65520/110909410-585ae780-8310-11eb-8844-634d48549085.png">

We may be able to tease out the missing "Compiling" lines later on but I think this is acceptable for now.

There's quite a bit of cleaning up I need to do but I think this is workable.

**Comment by finestructure** [comment]
For sake of completeness: compile failures are also correctly reported: https://staging.swiftpackageindex.com/builds/D2F5CB43-FFEA-4385-A3EE-1C4A6481C4C0

<img width="734" alt="CleanShot 2021-03-12 at 13 09 04@2x" src="https://user-images.githubusercontent.com/65520/110938435-278ca980-8334-11eb-807d-19852cd48978.png">

**Comment by finestructure** [comment]
Our first Linux based Linux builder is now online. Here are a few jobs it's processed:

- https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1098792893
- https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1098792944
- https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1098792989

However, we seem to having yet another problem. A few linux builds are failing when they shouldn't:

- https://staging.swiftpackageindex.com/builds/8565C3F1-B256-47DB-97B7-86853318D9B6
- https://staging.swiftpackageindex.com/builds/77555754-798D-469E-B378-C74A13D84C4B

The underlying issue seems to be:

```
<unknown>:0: error: input file 'CharacterSet+ext.swift' was modified during the build<unknown>:0: error: input file 'Match.swift' was modified during the build<unknown>:0: error: input file 'Parser+zip.swift' was modified during the build<unknown>:0: error: input file 'Parser.swift' was modified during the buildswiftc:
```

**Comment by finestructure** [comment]
I'm rolling back.

**Comment by finestructure** [comment]
Ok, I'm pretty sure the problem is that the `CI_RUNNER_ID` isn't unique per concurrent runner. I'm using that to create the docker volume but there's only one:

```
root@lb1:/home/gitlab-runner/builds/WmqMUs6e/0/finestructure# docker volume ls
DRIVER    VOLUME NAME
local     90a0457739bb9833eff0ad45704996c59de292aca357bac36e61f3de87f3b0a0
local     checkouts-4606859
local     de69f22b2942bd312da09b9da35980dd5bc39ce16d078167cbb32a84669c68c6
```

I'll switch to concurrency 1 and re-test. That should avoid us having to roll back until I've found a better solution.

**Comment by finestructure** [comment]
Indeed, there's a variable `CI_CONCURRENT_ID` that we should add to the volume id.

**Comment by finestructure** [comment]
Ok, all builds passing now - our first Linux on Linux builder is finally live and working :)

[Swift Package Index.pdf](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/files/6142018/Swift.Package.Index.pdf)

**Comment by finestructure** [comment]
Second runner is now online as well: https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/1098954836
